### PR TITLE
Added components like loss function and optimizers etc

### PR DESCRIPTION
**Pull request by gauravrele87**
## Description ##
Added a single fine that talks about loss function, autograd and optimizers

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+architectures include:
+
+- [Gluon CV model zoo](https://gluon-cv.mxnet.io/model_zoo/index.html)
+
+- [Gluon NLP model zoo](https://gluon-nlp.mxnet.io/model_zoo/index.html)
+
+Lets look at an example
+
+```{.python .input  n=117}
+from mxnet.gluon import model_zoo
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md:244]: The model_zoon in mxnet.gluon is no longer maintained. Users are recommended to use the GluonNLP, GluonCV packages

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+   "text/plain": "(1, 1000)"
+  },
+  "execution_count": 117,
+  "metadata": {},
+  "output_type": "execute_result"
+ }
+]
+```
+
+## Deciding the paradim for your NN
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md]:
```suggestion
## Deciding the paradigm for your NN
```

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+  },
+  "execution_count": 19,
+  "metadata": {},
+  "output_type": "execute_result"
+ }
+]
+```
+
+In the example above, the **-1** denotes that the size of the input layer is not
+specified during initialization.
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md:34]: Let's clarify that the size of the input layer will be inferred from the input data during the first forward pass

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+- `__init__` create the layers
+- `forward` define the forward function.
+
+```
+class Net(gluon.Block):
+    def __init__(self, **kwargs):
+        super(Net, self).__init__(**kwargs)
+
+    def forward(self, x):
+        pass
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md]:
```suggestion
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return x
```

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+```
+
+```{.python .input  n=54}
+class MLP(Block):
+    def __init__(self):
+        super(MLP, self).__init__()
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md]:
```suggestion
        super().__init__()
```

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+]
+```
+
+Simirly, you can the following code to implement a famous network called
+[LeNet](http://yann.lecun.com/exdb/lenet/) through `nn.Sequential`.
+
+```{.python .input  n=120}
+class LeNet(Block):
+    def __init__(self):
+        super(LeNet, self).__init__()
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md]:
```suggestion
        super().__init__()
```

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+
+1. Imperative API - Gluon that is very user friendly and allows for quick
+prototyping, easy debugging and natural control flow for people familiar with
+python programming.
+2. Symbolic or Declarative API for creating static graphs with low level
+optimizations on operators and allows for support in multiple languages, cross-
+compatibility with other frameworks (via ONNX) and as a backend for other high
+level APIs like keras. However, it’s less flexible because any logic must be
+encoded into the graph as special operators like scan, while_loop and cond. It’s
+also hard to debug.
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md]: Let's not refer to Imperative / Symbolic "API" here, as in MXNet 2 there is only the numpy compatible API. In MXNet 2 we can use the numpy compatible API in an imperative, eager fashion or we can use it in the lazy, deferred compute mode which traces / records all calls to the numpy operators. Based on the recording / trace a symbolic graph is constructed in the backend. The activation of lazy, deferred compute mode for recording / trace is automatically handled by HybridBlock.hybridize()

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+
+MXNet includes two types of programming paradigms:
+
+1. Imperative API - Gluon that is very user friendly and allows for quick
+prototyping, easy debugging and natural control flow for people familiar with
+python programming.
+2. Symbolic or Declarative API for creating static graphs with low level
+optimizations on operators and allows for support in multiple languages, cross-
+compatibility with other frameworks (via ONNX) and as a backend for other high
+level APIs like keras. However, it’s less flexible because any logic must be
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md]: Perhaps remove the "and as a backend for other high level APIs like keras"?

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+make 1000 forward passes through the network.
+
+```{.python .input  n=137}
+from time import time
+
+def bench(net, x):
+    start = time()
+    for i in range(1000):
+        y = net(x)
+    return time() - start
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md]: You need to do one forward pass before starting to record time, to ensure that deferred parameter initialization has already happened.

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+define a `HybridBlock`, we create the same `forward` function. MXNet takes care
+of hybridizing the model at the backend so you don't have to make changes to
+your code to convert it to a symbolic paradigm
+
+```{.python .input  n=152}
+from mxnet.gluon import HybridBlock
+
+class MLP_Hybrid(HybridBlock):
+    def __init__(self):
+        super(MLP_Hybrid, self).__init__()
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md]:
```suggestion
        super().__init__()
```

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+
+All of MXNet’s predefined layers are HybridBlocks. This means that any network
+consisting entirely of predefined MXNet layers can be compiled and run at much
+faster speeds by calling `.hybridize()`.
+
+## Next steps
+
+After you create a neural network, learn how to automatically
+compute the gradients in [Step 3: Automatic differentiation with
+autograd](3-autograd.md).
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md:486]: How does it relate to `3-components.md`?

File: docs/python_docs/python/tutorials/getting-started/crash-course/3-components.md [hunk 12]
```diff
@@ -0,0 +1,427 @@
...
+- classification: SigmoidBinaryCrossEntropyLoss, SoftmaxCrossEntropyLoss
+
+- embeddings: HingeLoss
+
+#### Customizing your Loss functions
+
+You can also create custom loss functions using **Loss Blocks**. For more
+information see []()
+
+You can inherit the base `Loss` class and write your own `hybrid_forward`
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/3-components.md]:
```suggestion
You can inherit the base `Loss` class and write your own `forward`
```

File: docs/python_docs/python/tutorials/getting-started/crash-course/3-components.md [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+When creating a **Gluon Trainer** you must provide the trainer object with
+1. A collection of parameters that need to be learnt. The collection of
+parameters will the weights and biases of your network that you are training.
+2. An Optimization algorithm (optimizer) that you want to use for training. This
+algorithm will be used to update the parameters every training iteration when
+`trainer.step` is called. For more information, see [optimizers in v1.6 TODO:
+CHANGE](https://mxnet.apache.org/versions/1.6/api/python/docs/api/optimizer/index.html)
+
+```{.python .input  n=17}
+curr_weight = net.weight.data().copy()
```
Comment (leezu) [inline docs/python_docs/python/tutorials/getting-started/crash-course/3-components.md]: Why do you call `.copy()` here?

**Review by leezu — COMMENTED** [review]
Thank you @gauravrele87. Would it make sense to delete the files superseded by this PR, ie "2-nn.md"?

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md (hunk 2 above)
Comment (gauravrele87) [reply docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md]: Fixed

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md (hunk 3 above)
Comment (gauravrele87) [reply docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md:34]: It was specified about 2 cells down below but i've put that in bold. I'll do another commit with that. Let me know if that works. For that specific example, i didnt want to add too much info than necessary so i thought i would add at that the end.

File: docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md (hunk 11 above)
Comment (gauravrele87) [reply docs/python_docs/python/tutorials/getting-started/crash-course/2-create-nn.md:486]: hmm we are still looking at what the right flow is so please disregard that section atm. I will update it once we know which way we are going.
